coap-gw: rework subscriptions

Danielius1922 opened the pull request.

Danielius1922 committed:
coap-gw: move observation to stand-alone file

Danielius1922 committed:
coap-gateway: refactor resource observations

Danielius1922 committed:
test: start two device simulators on make env

Start a device with /oic/res observation support disabled
(same device name as before) and a device with /oic/res
observation enabled.

Danielius1922 committed:
coap-gw: add function to query resource observability

Danielius1922 committed:
coap-gateway: rework /oic/res observation

Use Get method on coap connection to determine whether
/oic/res resource exists and is observable. Based on the
result create either a single /oic/res observation or
observations for all published resources.

Danielius1922 committed:
Add missing close calls in test connections

Danielius1922 committed:
resource-directory: fix data race on copied pointer

Danielius1922 committed:
coap-gateway: send multiple notification on batch observation

Danielius1922 committed:
coap-gateway: use default interface for get and observe notifications

jkralik commented on coap-gateway/coapconv/coapconv.go:
```diff
@@ -237,6 +251,56 @@ func NewNotifyResourceChangedRequest(resourceID *commands.ResourceId, connection
 	}
 }
 
+func NewNotifyResourceChangedRequests(deviceID, connectionID string, req *pool.Message) ([]*commands.NotifyResourceChangedRequest, error) {
```
rename function like `NewNotifyResourceChangedRequestsFromBatchResourceDiscovery`

jkralik commented on coap-gateway/coapconv/coapconv.go:
```diff
@@ -237,6 +251,56 @@ func NewNotifyResourceChangedRequest(resourceID *commands.ResourceId, connection
+		return false
+	}
+
+	var requests []*commands.NotifyResourceChangedRequest
```
preallocate array

jkralik commented on coap-gateway/service/observation/resourcesObserver.go:
```diff
@@ -0,0 +1,338 @@
+			if !observable {
+				return
+			}
+		}
```
is this check needed -> we always want to send data via `o.callbacks.OnObserveResource` whether it is observable/not observable.

Danielius1922 replied:
Yeah, I wanted to ask what to do here. `Observe` sends `Get` for non-observable resource. For non-observable `/oic/res` I want this to do nothing, no notification should be send and the observation as a whole will fall back to observing of published resources. And for published resources I should know whether the resource is observable or not. Only for observable resources will `coapConn.Observe` be called, non-observable call `coapConn.Get`. 

So if the detection of observability is reliable then this will work correctly. But if we want to be defensive then we can do it that for batch `/oic/res` it will do this and in other cases it will call `callbacks.OnObserveResource` (or maybe it should call `callbacks.OnGetResourceContent`? Since for non-observable resource it gets a `Get` response.)

Danielius1922 replied:
changed so this happens only for `/oic/res`

jkralik commented on coap-gateway/service/observation/resourcesObserver.go:
```diff
@@ -0,0 +1,338 @@
+type resourcesObserver struct {
+	lock      sync.Mutex
+	deviceID  string
+	resources map[int64]*observedResource // [instanceID]
```
i'm thinking to change it to array -> unpublish happens very rarely and mostly all resources. array consumes less memory than map.

Danielius1922 replied:
changed to sorted slice

jkralik commented on coap-gateway/service/observation/resourcesObserver.go:
```diff
@@ -0,0 +1,338 @@
+func (o *resourcesObserver) removeResourceLocked(instanceID int64) {
+	delete(o.resources, instanceID)
+}
+
```
do we need to have one line function ?

jkralik commented on coap-gateway/service/resourceDirectory.go:
```diff
@@ -61,46 +68,92 @@ func validatePublish(w wkRd) error {
 	w = fixTimeToLive(w)
 	for i, link := range w.Links {
 		w.Links[i].DeviceID = w.DeviceID
 		w.Links[i].Href = fixHref(link.Href)
-		w.Links[i].InstanceID = getInstanceID(link.Href)
+		w.Links[i].InstanceID = resource.GetInstanceID(link.Href)
```
could you reset all other properties to empty values ? the response will be much smaller.

Danielius1922 replied:
You mean on `w.Links[i]`? But the data from it is used to generate `commands.PublishResourceLinksRequest`, which has a `Resources` field and that takes all the data from the link. Are you sure the data is not needed? If I follow our code correctly the resources are then stored in `ResourceLinksSnapshotTaken` struct in RA. 

However, it seems that RA only uses `DeviceID`, `Href` and `ResourceTypes` fields, so perhaps we can indeed simplify it.

Danielius1922 committed:
test: skip tests for invalid device

Danielius1922 committed:
coap-gateway/observation: use sorted slice instead of map

Use sorted slice to keep track of observed resources in
resources observer to lower memory usage.

jkralik commented on coap-gateway/service/resourceDirectory.go:
```diff
@@ -110,7 +163,18 @@ func resourceDirectoryPublishHandler(req *mux.Message, client *Client) {
+	resp.Links = make(schema.ResourceLinks, len(w.Links))
+	for i, link := range w.Links {
+		resp.Links[i].Href = link.Href
+		resp.Links[i].InstanceID = link.InstanceID
```
why you don't send deviceID and anchor...
![image](https://user-images.githubusercontent.com/23374974/144106872-91599efb-249b-4ed0-83c0-51871f0226ef.png)

Danielius1922 replied:
Only Href and IstanceID are used by iotivity-lite, so that's the baseline that must be included. I wasn't sure about anything else. 

I'm not sure what does the spec tell me. It says that the payload should have the same information as the original UPDATE request. Doesn't that mean that we should include everything in response (as was done originally)?

Danielius1922 replied:
Reverted to original after SM discussion.

Danielius1922 committed:
Update plgd-dev/device go module

Danielius1922 committed:
coap-gateway: keep track of published resources for RA
